### ISSUE TITLE
timestampProcessPlugin not allow multiEntity plugin results in NGSIv2

### ISSUE THREAD

**Issue by AlvaroVega** [telefonicaid/iotagent-node-lib#621]
Related with PR about multientity and NGSIv2: https://github.com/dcalvoalonso/iotagent-node-lib/pull/5

If config.iota.timestamp flag is active then timestampProcessPlugin is executing before multiEntity plugin. Multientity plugin expands payload json in an array of entities. But TimeStampProcess plugin only expects a entity, and transformation performed by timestampProcess generates a wrong payload.

Here https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/services/ngsi/ngsiService.js#L435 result is not an array if timeStampPlugin is active.

Here https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/timestampProcessPlugin.js#L59 entity is an array of entites when multientity is preset

To reproduce it, add config.iota.timestamp flag to 
https://github.com/dcalvoalonso/iotagent-node-lib/pull/5/files#diff-8f552cb201d86830244748779fd2ee9bR35

**Comment by dcalvoalonso** [comment]
Sure @AlvaroVega! I will check it right now and provide a solution ASAP

**Comment by AlvaroVega** [comment]
Example:

$ curl -i -X GET 'http://localhost:7896/iot/d?i=disp1&d=t|18&k=vfd5f8n3ry0mk0vwnlr0niopw'

Produces a logs like:
```
iot-iota-ul                 | time=2018-06-12T10:33:40.787Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IOTAUL.HTTP.Binding | srv=smartcity | subsrv=/ | msg=Parsing payload [t|18] | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.787Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IOTAUL.HTTP.Binding | srv=smartcity | subsrv=/ | msg=Parsed data: [[{"t":"18"}]] | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.787Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.MongoDBGroupRegister | srv=smartcity | subsrv=/ | msg=Processing multiple HTTP measures for device [disp1] with apiKey ["vfd5f8n3ry0mk0vwnlr0niopw"] | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.788Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.MongoDBGroupRegister | srv=smartcity | subsrv=/ | msg=Looking for entity params ["resource","apikey"] with queryObj {"apikey":"vfd5f8n3ry0mk0vwnlr0niopw"} | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.791Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.MongoDBGroupRegister | srv=smartcity | subsrv=/ | msg=Device group data found: {"_id":"5ae181f455c5290001c16061","subservice":"/","service":"smartcity","type":"thin","apikey":"vfd5f8n3ry0mk0vwnlr0niopw","resource":"/iot/d","internalAttributes":[],"attributes":[],"lazy":[],"staticAttributes":[],"commands":[]} | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.792Z | lvl=ERROR | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.Alarms | srv=smartcity | subsrv=/ | msg=Releasing [MONGO-ALARM] | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.792Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.MongoDBDeviceRegister | srv=smartcity | subsrv=/ | msg=Looking for entity with id [disp1]. | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.797Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.MongoDBGroupRegister | srv=smartcity | subsrv=/ | msg=Looking for entity params ["resource","apikey"] with queryObj {"resource":"thin"} | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.799Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.MongoDBGroupRegister | srv=smartcity | subsrv=/ | msg=Device group for fields [["resource","apikey"]] not found: [{"resource":"thin"}] | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.800Z | lvl=ERROR | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.Alarms | srv=smartcity | subsrv=/ | msg=Raising [MONGO-ALARM]: undefined | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.805Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.NGSIService | srv=smartcity | subsrv=/ | msg=Updating device value in the Context Broker at [http://iot-orion:1026/v2/entities/thin:disp1/attrs] | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.806Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.NGSIService | srv=smartcity | subsrv=/ | msg=Using the following request:
iot-iota-ul                 | 
iot-iota-ul                 | {
iot-iota-ul                 |     "url": "http://iot-orion:1026/v2/entities/thin:disp1/attrs",
iot-iota-ul                 |     "method": "POST",
iot-iota-ul                 |     "headers": {
iot-iota-ul                 |         "fiware-service": "smartcity",
iot-iota-ul                 |         "fiware-servicepath": "/"
iot-iota-ul                 |     },
iot-iota-ul                 |     "json": {
iot-iota-ul                 |         "0": {
iot-iota-ul                 |             "type": "thin",
iot-iota-ul                 |             "metadata": {
iot-iota-ul                 |                 "TimeInstant": {
iot-iota-ul                 |                     "type": "DateTime",
iot-iota-ul                 |                     "value": "2018-06-12T10:33:40.805Z"
iot-iota-ul                 |                 }
iot-iota-ul                 |             }
iot-iota-ul                 |         },
iot-iota-ul                 |         "1": {
iot-iota-ul                 |             "type": "thin",
iot-iota-ul                 |             "metadata": {
iot-iota-ul                 |                 "TimeInstant": {
iot-iota-ul                 |                     "type": "DateTime",
iot-iota-ul                 |                     "value": "2018-06-12T10:33:40.805Z"
iot-iota-ul                 |                 }
iot-iota-ul                 |             }
iot-iota-ul                 |         },
iot-iota-ul                 |         "TimeInstant": {
iot-iota-ul                 |             "type": "DateTime",
iot-iota-ul                 |             "value": "2018-06-12T10:33:40.805Z"
iot-iota-ul                 |         }
iot-iota-ul                 |     }
iot-iota-ul                 | }
iot-iota-ul                 | 
iot-iota-ul                 |  | comp=IoTAgent
iot-iota-ul                 | time=2018-06-12T10:33:40.812Z | lvl=DEBUG | corr=0eaeb863-2453-4913-b8d1-dd7c716d04aa | trans=0eaeb863-2453-4913-b8d1-dd7c716d04aa | op=IoTAgentNGSI.NGSIService | srv=smartcity | subsrv=/ | msg=Received the following response from the CB: Value updated successfully
```

**Comment by dcalvoalonso** [comment]
A solution should be ready as part of PR #623. Could you please take a look at it?

Thanks in advance.

**Comment by fgalan** [comment]
Fixed by PR #623

**Comment by dcalvoalonso** [comment]
Hi @AlvaroVega,

I have done new tests and the problem was related to the fact that in https://github.com/telefonicaid/iotagent-node-lib/blob/master/lib/plugins/timestampProcessPlugin.js#L68, an error was raised if the IOTA was trying to update an entity but did not include any attribute. I have corrected this behaviour.

Moreover, I have also introduced some fixes in the rest of the library, mostly related to adding timestamps in operations involving the multientity plugin. This case was not included in the original unit tests.
